From 1ffa3ff5075fb47ed2e394ffc128662ce66ec49f Mon Sep 17 00:00:00 2001
From: Andrew Boie <andrew.p.boie@intel.com>
Date: Tue, 27 Nov 2012 12:58:19 -0800
Subject: init: HACK: Try multiple times to open console

If the console is set to be a USB-to-serial adapter, there is
a bootup race for the creation of its associated device nodes
and init's one-time console detection.

If the console can't be opened, try polling up to 5 seconds. Log
an error if it ultimately fails.

A better solution would be to reimplement the console functionality
to be hot-pluggable, but this would require much more invasive changes
to make it robust.

Change-Id: I20b19095a2f22efffafe2a7ac22d2be753a62d4b
Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
---
 init/init.c |   32 ++++++++++++++++++++------------
 1 files changed, 20 insertions(+), 12 deletions(-)

diff --git a/init/init.c b/init/init.c
index cd302e6..9c172ca 100755
--- a/init/init.c
+++ b/init/init.c
@@ -560,26 +560,33 @@ static int keychord_init_action(int nargs, char **args)
     return 0;
 }
 
+#define SLEEP_MS    50
+
 static int console_init_action(int nargs, char **args)
 {
     int fd;
     char tmp[PROP_VALUE_MAX];
+    int count = (5000 / SLEEP_MS);
 
     if (console[0]) {
         snprintf(tmp, sizeof(tmp), "/dev/%s", console);
         console_name = strdup(tmp);
     }
 
-    fd = open(console_name, O_RDWR);
-    if (fd >= 0)
-        have_console = 1;
-    close(fd);
+    load_565rle_image(INIT_IMAGE_FILE);
 
-    if( load_565rle_image(INIT_IMAGE_FILE) ) {
-        fd = open("/dev/tty0", O_WRONLY);
-        if (fd >= 0) {
-            const char *msg;
-                msg = "\n"
+    /* Block the boot until the console node comes up */
+    while (1) {
+        fd = open(console_name, O_WRONLY);
+        if (fd < 0 && count--)
+            usleep(SLEEP_MS * 1000);
+        else
+            break;
+    }
+    if (fd >= 0) {
+        have_console = 1;
+        const char *msg;
+        msg = "\n"
             "\n"
             "\n"
             "\n"
@@ -594,9 +601,10 @@ static int console_init_action(int nargs, char **args)
             "\n"
             "\n"
             "             A N D R O I D ";
-            write(fd, msg, strlen(msg));
-            close(fd);
-        }
+        write(fd, msg, strlen(msg));
+        close(fd);
+    } else {
+        ERROR("Gave up trying to open console %s\n", console_name);
     }
     return 0;
 }
-- 
1.7.4.1

